freeipa/src/main/java/com/sequenceiq/freeipa/kerberosmgmt/v1/KerberosMgmtV1Service.java

```diff
@@ -75,12 +81,16 @@ public ServiceKeytabResponse generateServiceKeytab(ServiceKeytabRequest request,
         Stack freeIpaStack = getFreeIpaStack(request.getEnvironmentCrn(), accountId);
         String realm = getRealm(freeIpaStack);
         ipaClient = freeIpaClientFactory.getFreeIpaClientForStack(freeIpaStack);
-        String adminUser = freeIpaClientFactory.getAdminUser();
         hostAdd(request.getServerHostName(), ipaClient);
-        String servicePrincipal = serviceAdd(request.getServiceName(), request.getServerHostName(), realm, adminUser, ipaClient);
-        String serviceKeytab = getServiceKeytab(servicePrincipal, ipaClient);
+        com.sequenceiq.freeipa.client.model.Service service = serviceAdd(request, realm, ipaClient);
+        String serviceKeytab;
+        if (service.getHasKeytab() && request.getDoNotRecreateKeytab()) {
+            serviceKeytab = getExistingServiceKeytab(service.getKrbprincipalname(), ipaClient);
+        } else {
+            serviceKeytab = getServiceKeytab(service.getKrbprincipalname(), ipaClient);
```

Comment:
The race condition with `doNotRecreateKeytab` that I mentioned earlier can cause this line to be executed more than once resulting in invalidation of the initial keytab.

Reply:
I think if here we can have a race condition, we can have it with 2 calls.
2 calls come to the get, both fails as there is no keytab. Both calls the create, the latter wins. I think using this 2 call solution won't solve the race condition itself. If we want to really solve it, then we have to block the read for the service if there is a create in operation for the same service, and drop new creates until there is a running create.